pelican/readers.py

```diff
@@ -630,8 +630,20 @@ def path_metadata(full_path, source_path, settings=None):
         if settings.get('DEFAULT_DATE', None) == 'fs':
             metadata['date'] = SafeDatetime.fromtimestamp(
                 os.stat(full_path).st_mtime)
-        metadata.update(settings.get('EXTRA_PATH_METADATA', {}).get(
-            source_path, {}))
+
+        # Pull EXTRA_PATH_METADATA for the source path and the paths of all its
+        # parents. Child paths are applied after parent paths; this way, if
+        # multiple EPM entries apply conflicting metadata to the same file, the
+        # most specific entry wins.
+
+        paths = []
+        path = source_path
+        while path:
+            paths.append(path)
+            path = os.path.split(path)[0]
+        epm = settings.get('EXTRA_PATH_METADATA', {})
+        for path in reversed(paths):
+            metadata.update(epm.get(path, {}))
```

Comment:
I'm wondering whether it is really necessary to break down the `source_path` into its ancestors like that, and whether it wouldn't be simpler to try to match the `epm` paths:

```
epm = [(p, m) for p, m in epm.items() if p == source_path or source_path.startswith(os.path.join(p, '')))
for _, m in sorted(epm):
    metadata.update(m)
```
In particular, this would automatically skip all comparison if `epm` is empty. (This version also tolerates trailing slashes, so we don't need to warn people against those.)

If you prefer your own version, could you at least add a check whether `epm` isn't empty?

Reply:
I think it was done this way to guarantee "if multiple EPM entries apply conflicting metadata to the same file, the most specific entry wins."

Reply:
I think avaris is right, but it's been a while so I don't remember for sure.

Reply:
Of course, but that's not the point I was trying to make, and my version should also let more specific paths overrule others (that's what `sorted` ensures). The following is probably clearer:

```
for p, m in sorted(epm.items()):
    if p == source_path or source_path.startswith(os.path.join(p, ''))):
        metadata.update(m)
```

Reply:
Hrm. That is clearer, and easier to follow than my version. What does os.path.join do with a blank string, exactly? I'm guessing it adds a slash if one isn't present. I worry that .startswith will produce false positives, but I'm having trouble coming up with a  path that would cause it.

I'll fiddle with it later and maybe update the PR.

Reply:
- yes, it adds a slash if one isn't present. It's a bit hacky, admittedly.
- I don't see how `startswith` could produce false positives with the final slash.

The main thing that got me thinking was that your version generates all the ancestors of `source_path` even if `epm` is empty, which is probably the case for 99% of installations. As long as you address that, I'm happy. :-)

Reply:
Is slash necessary? Can't think of a reason why `source_path.startswith(p)` isn't enough.

Reply:
Potentially you could have articles `foo/bar.md` and `foobar/baz.md`, then if you key extra metadata with `foo`, you only want it to affect the former, not the latter.

Reply:
Oh, right :).